lib/view_model/deserialization_error.rb

```diff
@@ -306,6 +308,99 @@ def initialize(detail, nodes = [])
         @detail = detail
         super(nodes)
       end
+
+      # Database constraint errors are pretty opaque and stringly typed. We can
+      # do our best to parse out what metadata we can from the error, and fall
+      # back when we can't.
+      def self.from_exception(exception, nodes = [])
+        case exception.cause
+        when PG::UniqueViolation
+          UniqueViolation.from_postgres_error(exception.cause, nodes)
+        else
+          self.new(exception.message, nodes)
+        end
+      end
+    end
+
+    class UniqueViolation < DeserializationError
+      status 400
+      attr_reader :detail, :constraint, :columns, :values
+
+      PG_ERROR_FIELD_CONSTRAINT_NAME = 'n'.ord # Not exposed in pg gem
+      def self.from_postgres_error(err, nodes)
+        result         = err.result
+        constraint     = result.error_field(PG_ERROR_FIELD_CONSTRAINT_NAME)
+        message_detail = result.error_field(PG::Result::PG_DIAG_MESSAGE_DETAIL)
+
+        columns, values = parse_message_detail(message_detail)
+
+        unless columns
+          # Couldn't parse the detail message, fall back on an unparsed error
+          return DatabaseConstraint.new(err.message, nodes)
+        end
+
+        self.new(err.message, constraint, columns, values, nodes)
+      end
+
+      class << self
+        DETAIL_PREFIX = 'Key ('
+        DETAIL_SUFFIX = ') already exists.'
+        DETAIL_INFIX  = ')=('
+        def parse_message_detail(detail)
+          stream = detail.dup
+
+          return nil unless stream.delete_prefix!(DETAIL_PREFIX)
+          return nil unless stream.delete_suffix!(DETAIL_SUFFIX)
+
+          # The message should start with an identifier list: pop off identifier
+          # tokens while we can.
+          identifiers = []
+
+          identifier = parse_identifier(stream)
```

Comment:
Awkward to have to unroll the loop body once due to the ordering of conditions. Would you prefer `loop/break`?